// semantic-ui-react/dist/commonjs/lib/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
exports.__esModule = true;
var _exportNames = {
  ModernAutoControlledComponent: true,
  eventStack: true,
  getComponentType: true,
  getUnhandledProps: true,
  isBrowser: true,
  doesNodeContainClick: true,
  leven: true,
  createPaginationItems: true,
  normalizeTransitionDuration: true,
  objectDiff: true,
  isRefObject: true,
  useAutoControlledValue: true,
  useClassNamesOnNode: true,
  useEventCallback: true,
  useForceUpdate: true,
  useIsomorphicLayoutEffect: true,
  useMergedRefs: true,
  usePrevious: true,
  useKeyOnly: true,
  useKeyOrValueAndKey: true,
  useValueAndKey: true,
  useMultipleProp: true,
  useTextAlignProp: true,
  useVerticalAlignProp: true,
  useWidthProp: true,
  htmlInputAttrs: true,
  htmlInputEvents: true,
  htmlInputProps: true,
  htmlImageProps: true,
  partitionHTMLProps: true,
  numberToWordMap: true,
  numberToWord: true,
  setRef: true,
  childrenUtils: true,
  customPropTypes: true,
  SUI: true
};
exports.useWidthProp = exports.useVerticalAlignProp = exports.useValueAndKey = exports.useTextAlignProp = exports.usePrevious = exports.useMultipleProp = exports.useMergedRefs = exports.useKeyOrValueAndKey = exports.useKeyOnly = exports.useIsomorphicLayoutEffect = exports.useForceUpdate = exports.useEventCallback = exports.useClassNamesOnNode = exports.useAutoControlledValue = exports.setRef = exports.partitionHTMLProps = exports.objectDiff = exports.numberToWordMap = exports.numberToWord = exports.normalizeTransitionDuration = exports.leven = exports.isRefObject = exports.isBrowser = exports.htmlInputProps = exports.htmlInputEvents = exports.htmlInputAttrs = exports.htmlImageProps = exports.getUnhandledProps = exports.getComponentType = exports.eventStack = exports.doesNodeContainClick = exports.customPropTypes = exports.createPaginationItems = exports.childrenUtils = exports.SUI = exports.ModernAutoControlledComponent = void 0;
var _ModernAutoControlledComponent = _interopRequireDefault(require("./ModernAutoControlledComponent"));
exports.ModernAutoControlledComponent = _ModernAutoControlledComponent.default;
var _childrenUtils = _interopRequireWildcard(require("./childrenUtils"));
exports.childrenUtils = _childrenUtils;
var _classNameBuilders = require("./classNameBuilders");
exports.useKeyOnly = _classNameBuilders.useKeyOnly;
exports.useKeyOrValueAndKey = _classNameBuilders.useKeyOrValueAndKey;
exports.useValueAndKey = _classNameBuilders.useValueAndKey;
exports.useMultipleProp = _classNameBuilders.useMultipleProp;
exports.useTextAlignProp = _classNameBuilders.useTextAlignProp;
exports.useVerticalAlignProp = _classNameBuilders.useVerticalAlignProp;
exports.useWidthProp = _classNameBuilders.useWidthProp;
var _customPropTypes = _interopRequireWildcard(require("./customPropTypes"));
exports.customPropTypes = _customPropTypes;
var _eventStack = _interopRequireDefault(require("./eventStack"));
exports.eventStack = _eventStack.default;
var _factories = require("./factories");
Object.keys(_factories).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _factories[key]) return;
  exports[key] = _factories[key];
});
var _getComponentType = _interopRequireDefault(require("./getComponentType"));
exports.getComponentType = _getComponentType.default;
var _getUnhandledProps = _interopRequireDefault(require("./getUnhandledProps"));
exports.getUnhandledProps = _getUnhandledProps.default;
var _htmlPropsUtils = require("./htmlPropsUtils");
exports.htmlInputAttrs = _htmlPropsUtils.htmlInputAttrs;
exports.htmlInputEvents = _htmlPropsUtils.htmlInputEvents;
exports.htmlInputProps = _htmlPropsUtils.htmlInputProps;
exports.htmlImageProps = _htmlPropsUtils.htmlImageProps;
exports.partitionHTMLProps = _htmlPropsUtils.partitionHTMLProps;
var _isBrowser = _interopRequireDefault(require("./isBrowser"));
exports.isBrowser = _isBrowser.default;
var _doesNodeContainClick = _interopRequireDefault(require("./doesNodeContainClick"));
exports.doesNodeContainClick = _doesNodeContainClick.default;
var _leven = _interopRequireDefault(require("./leven"));
exports.leven = _leven.default;
var _createPaginationItems = _interopRequireDefault(require("./createPaginationItems"));
exports.createPaginationItems = _createPaginationItems.default;
var _SUI = _interopRequireWildcard(require("./SUI"));
exports.SUI = _SUI;
var _numberToWord = require("./numberToWord");
exports.numberToWordMap = _numberToWord.numberToWordMap;
exports.numberToWord = _numberToWord.numberToWord;
var _normalizeTransitionDuration = _interopRequireDefault(require("./normalizeTransitionDuration"));
exports.normalizeTransitionDuration = _normalizeTransitionDuration.default;
var _objectDiff = _interopRequireDefault(require("./objectDiff"));
exports.objectDiff = _objectDiff.default;
var _isRefObject = _interopRequireDefault(require("./isRefObject"));
exports.isRefObject = _isRefObject.default;
var _useAutoControlledValue = _interopRequireDefault(require("./hooks/useAutoControlledValue"));
exports.useAutoControlledValue = _useAutoControlledValue.default;
var _useClassNamesOnNode = _interopRequireDefault(require("./hooks/useClassNamesOnNode"));
exports.useClassNamesOnNode = _useClassNamesOnNode.default;
var _useEventCallback = _interopRequireDefault(require("./hooks/useEventCallback"));
exports.useEventCallback = _useEventCallback.default;
var _useForceUpdate = _interopRequireDefault(require("./hooks/useForceUpdate"));
exports.useForceUpdate = _useForceUpdate.default;
var _useIsomorphicLayoutEffect = _interopRequireDefault(require("./hooks/useIsomorphicLayoutEffect"));
exports.useIsomorphicLayoutEffect = _useIsomorphicLayoutEffect.default;
var _useMergedRefs = _interopRequireWildcard(require("./hooks/useMergedRefs"));
exports.useMergedRefs = _useMergedRefs.default;
exports.setRef = _useMergedRefs.setRef;
var _usePrevious = _interopRequireDefault(require("./hooks/usePrevious"));
exports.usePrevious = _usePrevious.default;
