// semantic-ui-react/dist/commonjs/views/Comment/CommentAvatar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
exports.__esModule = true;
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var React = _interopRequireWildcard(require("react"));
var _lib = require("../../lib");
/**
 * A comment can contain an image or avatar.
 */
var CommentAvatar = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var className = props.className,
    src = props.src;
  var classes = (0, _clsx.default)('avatar', className);
  var rest = (0, _lib.getUnhandledProps)(CommentAvatar, props);
  var _partitionHTMLProps = (0, _lib.partitionHTMLProps)(rest, {
      htmlProps: _lib.htmlImageProps
    }),
    imageProps = _partitionHTMLProps[0],
    rootProps = _partitionHTMLProps[1];
  var ElementType = (0, _lib.getComponentType)(props);
  return /*#__PURE__*/React.createElement(ElementType, (0, _extends2.default)({}, rootProps, {
    className: classes,
    ref: ref
  }), (0, _lib.createHTMLImage)(src, {
    autoGenerateKey: false,
    defaultProps: imageProps
  }));
});
CommentAvatar.handledProps = ["as", "className", "src"];
CommentAvatar.displayName = 'CommentAvatar';
CommentAvatar.propTypes = process.env.NODE_ENV !== "production" ? {
  /** An element type to render as (string or function). */
  as: _propTypes.default.elementType,
  /** Additional classes. */
  className: _propTypes.default.string,
  /** Specifies the URL of the image. */
  src: _propTypes.default.string
} : {};
var _default = exports.default = CommentAvatar;
